File: Data Structures/Java/Depth First Search.java
import java.util.Stack;

class Graph {
    private int vertices;
    private int[][] adjacencyMatrix;

    public Graph(int vertices) {
        this.vertices = vertices;
        this.adjacencyMatrix = new int[vertices][vertices];
    }

    public void addEdge(int src, int dest) {
        this.adjacencyMatrix[src][dest] = 1;
        this.adjacencyMatrix[dest][src] = 1;
    }

    public void DFS(int startVertex) {
        boolean[] visited = new boolean[vertices];
        Stack<Integer> stack = new Stack<>();

        visited[startVertex] = true;
        System.out.print(startVertex + " ");

        stack.push(startVertex);

        while (!stack.isEmpty()) {
            int currentVertex = stack.pop();

            for (int i = 0; i < vertices; ++i) {
                if (adjacencyMatrix[currentVertex][i] == 1 && !visited[i]) {
                    visited[i] = true;
                    System.out.print(i + " ");
                    stack.push(i);
                }
            }
        }
    }

    public static void main(String[] args) {
        int vertices = 5;
        Graph graph = new Graph(vertices);

        graph.addEdge(0, 1);
        graph.addEdge(0, 2);
        graph.addEdge(1, 3);
        graph.addEdge(1, 4);

        System.out.print("Depth-First Traversal: ");
        graph.DFS(0);
    }
}

File: Data Structures/Python/Depth First Search.py
from collections import defaultdict

class Graph:
    def __init__(self, vertices):
        self.vertices = vertices
        self.adjacency_matrix = defaultdict(list)

    def add_edge(self, src, dest):
        self.adjacency_matrix[src].append(dest)
        self.adjacency_matrix[dest].append(src)

    def DFS(self, start_vertex):
        visited = [False] * self.vertices
        stack = [start_vertex]

        visited[start_vertex] = True
        print(start_vertex, end=" ")

        while stack:
            current_vertex = stack.pop()

            for neighbor in self.adjacency_matrix[current_vertex]:
                if not visited[neighbor]:
                    visited[neighbor] = True
                    print(neighbor, end=" ")
                    stack.append(neighbor)

if __name__ == "__main__":
    vertices = 5
    graph = Graph(vertices)

    graph.add_edge(0, 1)
    graph.add_edge(0, 2)
    graph.add_edge(1, 3)
    graph.add_edge(1, 4)

    print("Depth-First Traversal:", end=" ")
    graph.DFS(0)
